_posts/2021-11-04-prefix-sum-portable.md

```diff
@@ -0,0 +1,147 @@
+
+The fact that FXC is on the critical path points to a deeper difference in approach between WebGPU and piet-gpu. By necessity, shaders pulled from the Web must be compiled at runtime, and FXC is considerably more convenient for that than the more advanced [DXC], as FXC is included in Windows and can be called by any application. The DLL for DXC is approximately 20MB.
+
+In piet-gpu, by contrast, shaders are compiled to IR ahead of time as part of the build pipeline, so there is no need for runtime translation. That helps a lot with binary size and startup time, as well as a lot more flexibility in tools that can be run as part of the build.
+
+The Windows story with Dawn (Chrome's WebGPU implementation) continues to evolve. Currently they run with their own version of FXC, which brings compatibility to the widest range of devices (there's still a large fraction that can't run code from DXC) and lets them roll fixes if the version that ships with the OS is stale. I'm concerned that even so, there will be shaders that fail because of FXC limitations. There are a number of ways forward, including direct translation from WGSL into DXBC (and DXIL as well, for access to [Shader Model 6] capabilities), but at best this work will take a while.
+
+There is a [WebGPU implemenation of decoupled look-back] by Reese Levine, but at this point it's not expected to pass. Hopefully it will be useful to shake out bugs (it would pass if the barrier scope were upgraded), and also to point the way for extensions to WebGPU that would allow this algorithm to run.
+
+### Shader translation
+
+Another concern I have about WebGPU is that it is difficult to maintain shaders as single authoritative source files and translate them automatically into WGSL. For all other shader languages, we've arrived at a reasonable solution involving GLSL source and translation using spirv-tools to HLSL and Metal (enabling translation to intermediate representation as well). That won't work for WGSL, and, not surprisingly, the problem is atomics.
+
+The fundamental issue is that [atomics][Atomics proposal] in WGSL are strictly typed, following the precedent of Metal, which in turn follows C++. By contrast, in HLSL and GLSL, atomics are operations that can work on ordinary memory locations. Arguably, that's not ideal, but it is how shaders are written today. That doesn't create a portability problem in practice, because Metal also supports unsafe C++ casting operations, and this pattern is widely used (by spirv-cross and perhaps ironically also [Tint]) to compile the standard atomics of shader languages into Metal.
```

Comment:
I believe code that exercises unsafe casting supported by Metal tooling will lead to undefined behaviour, because it violates the TBAA rules.

Specifically, see the comment [here](https://github.com/gpuweb/gpuweb/issues/1931#issuecomment-953180418). In that case we concluded that Metal's shading language doesn't support operating on a pointer-to-vector-component. For example, casting a pointer-to-float4 to pointer-to-float, then writing the float, will yield undefined behaviour if you try to read back the result via a pointer-to-float4.

The same argument applies to disallow aliasing an atomic type with a regular scalar type. For example, by my reading of C++, `std::atomic<uint32_t>` is not an "aggregate type" (it has a "user-provided" constructor, for one thing.)
And so you can't write via a pointer-to-uint32_t and expect to read via atomic-load (and vice versa).

So even though you can "cheat" and write code that gets past the tools, and probably functions ok today, you're going outside the "spec" of C++ (by way of Metal Shading Language).

Reply:
This sounds like a genuine controversy. My perspective is that the pattern is a de facto standard in Metal, because it is the translation output from spirv-cross, regardless of how much they want to support it.

If you do want a strict interpretation of the Metal spec, then the inexorable conclusion is that you can't reliably run *any* workload involving atomics that's been automatically translated from GLSL or MLSL. Nor is it feasible in any practical way to automatically translate MSL to other shading languages. Thus, someone in my shoes has few good choices; to do it the "right" way requires writing custom versions for each of the popular shading languages.

I've reread that passage in the blog, and feel like I can stand by it. I am happy to change or clarify if you can suggest something, but I still feel like it fairly well communicates the current state of affairs to the intended audience.

Don't get me wrong, avoiding undefined behavior is a goal I agree with completely, but it has to be balanced with other concerns. Almost none of the stuff run in production today is based on an actual spec, though I hope that changes. In this case, I believe there's a strong risk that being overly strict will overall hinder progress toward that goal, rather than help.

Perhaps there's a better forum for this discussion? I intended https://github.com/gpuweb/gpuweb/issues/2229 to open this up, and there's been discussion of some of the issues there but not much on this one.

Reply:
Thinking about this some more, how is this situation any different than signed arithmetic overflow, which is carefully defined in SPIR-V and UB in Metal? The goals in 1.1 stated of the WGSL spec give clear guidance how to resolve that one. Why would atomics-as-operations be any different?